v3/MmlTree/ts/Node.ts

```diff
@@ -0,0 +1,355 @@
+     * @override
+     */
+    public setProperty(name: string, value: Property) {
+        this.properties[name] = value;
+    }
+
+    /*
+     * @override
+     */
+    public getProperty(name: string) {
+        return this.properties[name];
+    }
+
+    /*
+     * @override
+     */
+    public getPropertyNames() {
+        return Object.keys(this.properties);
+    }
+
+    /*
+     * @override
+     */
+    public getAllProperties() {
+        return this.properties;
+    }
+
+    /*
+     * @override
+     */
+    public removeProperty(...names: string[]) {
+        for (const name of names) {
+            delete this.properties[name];
+        }
+    }
+
+
+    /*
+     * @override
+     */
+    public isKind(kind: string): boolean {
+        return this.factory.nodeIsKind(this, kind);
+    }
+
+
+    /*
+     * @override
+     */
+    public setChildren(children: Node[]) {
+        if (this.childNodes.length) {
+            this.childNodes = [];
+        }
+        for (let child of children) {
+            this.appendChild(child);
+        }
+    }
+
+    /*
+     * @override
+     */
+    public appendChild(child: Node) {
+        this.childNodes.push(child);
+        child.parent = this;
+        return child;
+    }
+
+    /*
+     * @override
+     */
+    public replaceChild(newChild: Node, oldChild: Node) {
+        let i = this.childIndex(oldChild);
+        this.childNodes[i] = newChild;
+        newChild.parent = this;
+        return newChild;
+    }
+
+
+    /*
+     * @override
+     */
+    public childIndex(node: Node) {
+        let i = 0;
+        for (const child of this.childNodes) {
+            if (child === node) {
+                return i;
+            }
+            i++;
+        }
+        return null;
+    }
+
+
+    /*
+     * @override
+     */
+    public findNodes(kind: string) {
+        let nodes: Node[] = [];
+        this.walkTree((node: Node) => {
+            if (node.isKind(kind)) {
+                nodes.push(node);
+            }
+        });
+        return nodes;
+    }
+
+
+    /*
+     * @override
+     */
+    public walkTree(func: (node: Node, data?: any) => void, data?: any) {
+        func(this, data);
+        for (const child of this.childNodes) {
+            child.walkTree(func, data);
+        }
+        return data;
+    }
+
+}
+
+/*********************************************************/
+/*
+ *  The abstract EmptyNode class
+ */
+
+export abstract class AEmptyNode extends ANode {
+    /*
+     *  We don't have children, so ignore these methods
+     */
+
+    /*
+     * @override
+     */
+    public setChildren(children: Node[]) {
+    }
+
+    /*
+     * @override
+     */
+    public appendChild(child: Node) {
+        return child;
+    }
+
+    /*
+     * @override
+     */
+    public replaceChild(newChild: Node, oldChild: Node) {
+        return oldChild;
+    }
+
+    /*
+     * @override
+     */
+    public childIndex(node: Node) {
+        return null as number;
```

Comment:
Yes.  Otherwise it gets a return type of `any`, and typescript (with our configuration) complains.